lib/Sema/MiscDiagnostics.cpp

```diff
@@ -4438,6 +4438,137 @@ static void diagnoseExplicitUseOfLazyVariableStorage(const Expr *E,
   const_cast<Expr *>(E)->walk(Walker);
 }
 
+static void diagnoseComparisonWithNaN(const Expr *E, const DeclContext *DC) {
+  class ComparisonWithNaNFinder : public ASTWalker {
+    const ASTContext &C;
+    const DeclContext *DC;
+
+  public:
+    ComparisonWithNaNFinder(const DeclContext *dc)
+        : C(dc->getASTContext()), DC(dc) {}
+
+    void tryDiagnoseComparisonWithNaN(BinaryExpr *BE) {
+      ValueDecl *comparisonDecl = nullptr;
+
+      // The == and != methods take two arguments.
+      if (BE->getArg()->getNumElements() != 2) {
+        return;
+      }
+
+      // Dig out the function the arguments are being passed to.
+      if (auto Fn = BE->getFn()) {
+        if (auto DSCE = dyn_cast<DotSyntaxCallExpr>(Fn)) {
+          comparisonDecl = DSCE->getCalledValue();
+        } else {
+          comparisonDecl = BE->getCalledValue();
+        }
+      }
+
+      // Bail out if it isn't a function.
+      if (!comparisonDecl || !isa<FuncDecl>(comparisonDecl)) {
+        return;
+      }
+
+      // We're only interested in == and != functions.
+      auto comparisonDeclName = comparisonDecl->getBaseIdentifier();
+      if (!(comparisonDeclName.is("==") || comparisonDeclName.is("!=") ||
+            comparisonDeclName.is("<=") || comparisonDeclName.is("<") ||
+            comparisonDeclName.is(">") || comparisonDeclName.is(">="))) {
+        return;
+      }
+
+      auto firstArg = BE->getArg()->getElement(0);
+      auto secondArg = BE->getArg()->getElement(1);
+
+      // Both arguments must conform to FloatingPoint protocol.
+      auto conformsToFpProto = [&](Type type) {
+        auto fpProto = C.getProtocol(KnownProtocolKind::FloatingPoint);
+        return !TypeChecker::conformsToProtocol(type, fpProto,
+                                                const_cast<DeclContext *>(DC))
+                    .isInvalid();
+      };
+
+      if (!conformsToFpProto(firstArg->getType()) ||
+          !conformsToFpProto(firstArg->getType())) {
+        return;
+      }
+
+      // Dig out the declarations for the arguments.
+      ValueDecl *firstVal = nullptr;
+      ValueDecl *secondVal = nullptr;
+      if (auto DRE = dyn_cast<DeclRefExpr>(firstArg)) {
+        firstVal = DRE->getDecl();
+      } else if (auto MRE = dyn_cast<MemberRefExpr>(firstArg)) {
+        firstVal = MRE->getMember().getDecl();
+      }
+
+      if (auto DRE = dyn_cast<DeclRefExpr>(secondArg)) {
+        secondVal = DRE->getDecl();
+      } else if (auto MRE = dyn_cast<MemberRefExpr>(secondArg)) {
+        secondVal = MRE->getMember().getDecl();
+      }
+
+      // One of them has to be '.nan', so if we don't have declarations
+      // for both, then bail out.
```

Comment:
Extremely minor nit: should be "if we don't have declarations for _either_, then bail out," I think.